### PR TITLE
Automatically disconnect from voice channel

### LINKED ISSUE #8: Bot does not disconnect on sound completion
When the bot has finished playing the logs say
```
2021-10-12 20:43:33,191 - discord.player - INFO - Preparing to terminate ffmpeg process 361527.
2021-10-12 20:43:33,191 - discord.player - INFO - ffmpeg process 361527 successfully terminated with return code of 0.
```
however, the bot does not disconnect and stays in the voice channel.

### PR DESCRIPTION
**Pull request by michael-je**
Makes it so that the bot will automatically disconnect from the voice channel after a set amount of time of voice inactivity.

Also fixes a bug where that would happen when online media was not successfully found. Bot now correctly reports the error and ignores it.

Closes #8

### CODE REVIEW

File: bot.py
Comment (morgaesis) [inline bot.py:214]: Should the latest command time be registered if the command is not voice related? Should `!joke` reset the command timer?

File: bot.py
Comment (morgaesis) [inline bot.py:307]: Can we increase this to at least 1 minute?

File: bot.py
Comment (morgaesis) [inline bot.py:325]: Please move all non-critical sections out of the lock

File: bot.py
Comment (morgaesis) [inline bot.py:329]: Missing check for end-of-queue disconnect. I think the bot  `is_playing() == False` at the end of queue, so it will never disconnect upon reaching end-of-queue.

**Review by morgaesis — COMMENTED** [review]
Sweet code ☺️

File: bot.py
Comment (michael-je) [reply bot.py:214]: I think it should, yes. I would find it weird if the bot would randomly disconnect while it was doing somethinge else, even if it wasn't voice related.

Comment (michael-je) [reply bot.py:214]: Also, I want to try and minimize how often the users have to hear the disconnect chime in discord, while still keeping the bot feeling responzive.

File: bot.py
Comment (michael-je) [reply bot.py:329]: is it possible that you misread
```python
if self.voice_client and not self.voice_client.is_playing():
```
as
```python
if self.voice_client and self.voice_client.is_playing():
```
?

File: bot.py
Comment (gardarandri) [reply bot.py:307]: This could also be done without a loop. You could just schedule a disconnect in the after_callback:

```
async def attempt_disconnect():
     await asyncio.sleep(10min)
     if is_playing():
         # Something is playing meaning another disconnect will be scheduled once that is done.
         return
     disconnect()
     # Or check if enough inactivity has passed. For example by setting self.last_command_time in the after_callback function.
     if something something:
         disconnect()

def after_callback(f):
     ....
     asyncio.create_task(self.attempt_disconnect())
```

File: bot.py
Comment (morgaesis) [inline bot.py:313]:
```suggestion
        if not self.voice_client or self.voice_client.is_playing():
            return
        # ...rest of code
```

You could make the code slightly more readable by reducing nesting.

**Review by morgaesis — CHANGES_REQUESTED** [review]

File: bot.py
Comment (morgaesis) [reply bot.py:329]: Yes, sorry.

File: bot.py
Comment (michael-je) [reply bot.py:313]: nice!

File: bot.py
Comment (michael-je) [reply bot.py:307]: This is a much better solution, nice!

I was trying to do something similar a while ago but kept having a really hard time accessing the event loop. I came to the conclusion that it was impossible to do so directly because it was being handled by the discord.py library or something. I was just being dumb though and trying to do:
```python
asyncio.create_task()
```
instead of:
```python
self.loop.create_task()
```

The only issue I see with your solution is that if the last `attempt_disconnect()` does it's check before the `DISCONNECT_TIMER` is up, then the bot will potentially never disconnect, due to there being nothing to call a new `attempt_disconnect()`.

However, this can be easily worked around by just not caring about commands and instead only checking for voice activity. And with this solution we can actually do that very easily, so if you guys agree I'll implement this instead.

Comment (morgaesis) [reply bot.py:307]: Yes, only implement for voice-related commands.

With the *sleep-then-attempt-disconnect* route in `after_callback`, you only need the sleep timer. If the timer expires while something is playing, it shouldn't disconnect anyway. If the timer expires right after the next song finishes, you could check a (semi-global) token set right before sleeping; if the token is still the same, then the current sleep is the only sleep, and therefore is the *after-queue* sleep.

So something like this:
```python
async def attempt_disconnect():
    self.sleep_token = random.random()
    local_sleep_token = self.sleep_token
    await asyncio.sleep(self.SLEEP_TIMER)
    if self.sleep_token != local_sleep_token:
        # Not the latest disconnect attempt
        return
```

Comment (michael-je) [reply bot.py:307]: Yes that's what I was thinking as well ;)

File: bot.py
Comment (morgaesis) [inline bot.py:321]:
```suggestion
        if abs(time_since_last_played - self.DISCONNECT_TIMER_SECONDS) < buffer :
```
Allows timer drift in either direction

File: bot.py
Comment (morgaesis) [inline bot.py:431]: Swap these two lines, so the reaction comes immediately, not 5 minutes later.
```suggestion
        await message.add_reaction(MusicBot.REACTION_EMOJI)
        await self.attempt_disconnect()
```

**Review by morgaesis — APPROVED** [review]
LGTM :+1:

Two minor improvements, but I'm approving so as to not block merge once they are fixed. They aren't critical, anyway.

File: bot.py
Comment (gardarandri) [inline bot.py:11]: I would prefer 'import time' and then do time.time(). Otherwise you cannot use anything else from the time package.

File: bot.py
Comment (gardarandri) [reply bot.py:321]: Can this even happen?

File: bot.py
Comment (gardarandri) [inline bot.py:410]: Why not `self.loop.create_task(self.attempt_disconnect())`?

File: bot.py
Comment (gardarandri) [inline bot.py:430]: Why not `self.loop.create_task(self.attempt_disconnect())`?

**Review by gardarandri — APPROVED** [review]
LGTM modulo comments.

Thanks for implementing my suggestion!

**Comment by morgaesis** [comment]
@michael-je The bot does not disconnect if it is alone for a long time! Filed a new issue #50

File: bot.py
Comment (michael-je) [reply bot.py:11]: done

File: bot.py
Comment (michael-je) [reply bot.py:410]: good point

File: bot.py
Comment (michael-je) [reply bot.py:431]: fixed

File: bot.py
Comment (michael-je) [reply bot.py:321]: That's a good point gardarandri. Fixed.

**Pull request merged**
